### PR TITLE
Automatic download

### LINKED ISSUE #131: Create F9 and Ctrl-R keyboard shortcuts to Tokenize and Download
Create F9 and Ctrl-R keyboard shortcuts to activate the Download function, which causes the code to be tokenized and downloaded automatically (unless there's an error) via the operations described by issues #138 and #139.

### LINKED ISSUE #163: Automatic Download - Handle "No Modules Found"
**Scenario**
No modules of any kind are found to be connected

**Behavior**
Leave progress display on-screen (for user review)
Display error: "No BASIC Stamps found."
NOTE: the intent is for the user to be able to read the error and view the list of ports scanned to help them determined what was done and what the results were

Split out from #139

### LINKED ISSUE #164: Automatic Download - Single instance of module type found
**Scenario**
Only one instance of the source code's specified module type is found to be connected to the system

**Behavior**
Automatically select that module type's port and perform the entire programming protocol on it. This means reset, identify, program, verify.  Yes, the module will actually have been identified twice and programmed once through the processes of issues #138 and #139.

Split out from #139

### LINKED ISSUE #165: Automatic Download - Multiple instances of module type found
**Scenario**
More than one instance of the source code's specified module type is found to be connected to the system

**Behavior**
- Prompt the user to select the module they wish to download to; "Please select which module to download to."
- If the user selects the Cancel button; close the display and cancel operation
- If the user selects a module, immediately perform the entire programming protocol on it. This means reset, identify, program, verify. Yes, the module will actually have been identified twice and programmed once through the processes of issues #138 and #139.
  - NOTE: It's not desired, but if there must be a Download button that they have to click after selecting the desired module, it's acceptable. We'd like to avoid forcing this extra click, however.

Split out from #139

### PR DESCRIPTION
**Pull request by spatializes**
#### What's this PR do?

Adds automatic download functionality and logic to decide when to download. Address the issues linked below to deal with:
- No modules found
- 1 matching module found
- multiple matching modules found

Also adds a keypress for CTRL-R or F9 to automatically download code.
#### Where should the reviewer start?

Pull down this branch.
Get repo bs2-serial and pull down branch 'compiler-changes'
Get repo irken and pull down branch 'scan-options'
npm link to both repos
npm run build
#### How should this be manually tested?

Add some working code to your editor. With the correct directive and 1 matching STAMP plugged in, click the download button. The download window should pop up and automatically start downloading to the STAMP.

Try unplugging the STAMP. Download again, you should receive an error message as shown in the image below.

Try plugging in two STAMPS. You should get a message to pick one. Click a STAMP, it should automatically download without need for a Download button.

Try the CTRL-I and CTRL-R shortcut keys.
CTRL-I should Identify but not automatically download.
CTRL-R should bring up the download overlay and automatically download if a matching STAMP is found.
#### What are the relevant tickets?

Closes #131
Closes #163 
Closes #164 
Closes #165 
#### Screenshots

image

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -15,6 +15,7 @@
     "iggins": "^0.2.1",
     "irken": "^0.6.1",
     "lodash": "^3.9.1",
+    "pbasic-tokenizer": "^0.3.0",
```
Comment (phated) [inline package.json]: not needed anymore

File: plugins/overlays/download.js [hunk 2]
```diff
@@ -35,10 +34,23 @@ class DownloadOverlay extends React.Component {
       );
   }
 
+  componentizeUserMessage(message){
```
Comment (phated) [inline plugins/overlays/download.js]: doesn't use `this` so it shouldn't be attached to the class, just a normal function

File: plugins/overlays/download.js [hunk 3]
```diff
@@ -79,12 +93,6 @@ class DownloadOverlay extends React.Component {
     );
   }
 
-  _onAccept(){
-    const { handleSuccess, handleError, handleComplete } = this.props;
```
Comment (phated) [inline plugins/overlays/download.js:83]: if these are not used, they need to be removed from being passed into the overlay component

**Comment by spatializes** [comment]
updated.

**Comment by spatializes** [comment]
updated.

**Pull request merged**